scipy/optimize/least_squares.py

```diff
@@ -0,0 +1,819 @@
+            fun, x0, (), full_output, ftol, xtol, gtol,
+            max_nfev, epsfcn, factor, scaling)
+    else:
+        if max_nfev is None:
+            max_nfev = 100 * n
+        x, info, status = _minpack._lmder(
+            fun, jac, x0, (), full_output, col_deriv,
+            ftol, xtol, gtol, max_nfev, factor, scaling)
+
+    f = info['fvec']
+
+    if callable(jac):
+        J = jac(x)
+    else:
+        J = np.atleast_2d(approx_derivative(fun, x))
+
+    cost = 0.5 * np.dot(f, f)
+    g = J.T.dot(f)
+    g_norm = norm(g, ord=np.inf)
+
+    nfev = info['nfev']
+    njev = info.get('njev', None)
+
+    status = FROM_MINPACK_TO_COMMON[status]
+    active_mask = np.zeros_like(x0, dtype=int)
+
+    return OptimizeResult(
+        x=x, cost=cost, fun=f, jac=J, grad=g, optimality=g_norm,
+        active_mask=active_mask, nfev=nfev, njev=njev, status=status)
+
+
+def prepare_bounds(bounds, x0):
+    lb, ub = [np.asarray(b, dtype=float) for b in bounds]
+    if lb.ndim == 0:
+        lb = np.resize(lb, x0.shape)
+
+    if ub.ndim == 0:
+        ub = np.resize(ub, x0.shape)
+
+    return lb, ub
+
+
+def check_tolerance(ftol, xtol, gtol):
+    message = "{} is too low, setting to machine epsilon {}."
+    if ftol < EPS:
+        warn(message.format("`ftol`", EPS))
+        ftol = EPS
+    if xtol < EPS:
+        warn(message.format("`xtol`", EPS))
+        xtol = EPS
+    if gtol < EPS:
+        warn(message.format("`gtol`", EPS))
+        gtol = EPS
+
+    return ftol, xtol, gtol
+
+
+def check_scaling(scaling, x0):
+    if scaling == 'jac':
+        return scaling
+
+    try:
+        scaling = np.asarray(scaling, dtype=float)
+        valid = np.all(np.isfinite(scaling)) and np.all(scaling > 0)
+    except (ValueError, TypeError):
+        valid = False
+
+    if not valid:
+        raise ValueError("`scaling` must be 'jac' or array_like with "
+                         "positive numbers.")
+
+    if scaling.ndim == 0:
+        scaling = np.resize(scaling, x0.shape)
+
+    if scaling.shape != x0.shape:
+        raise ValueError("Inconsistent shapes between `scaling` and `x0`.")
+
+    return scaling
+
+
+def check_jac_sparsity(jac_sparsity, m, n):
+    if jac_sparsity is None:
+        return None
+
+    if not issparse(jac_sparsity):
+        jac_sparsity = np.atleast_2d(jac_sparsity)
+
+    if jac_sparsity.shape != (m, n):
+        raise ValueError("`jac_sparsity` has wrong shape.")
+
+    return jac_sparsity, group_columns(jac_sparsity)
+
+
+def least_squares(
+        fun, x0, jac='2-point', bounds=(-np.inf, np.inf), method='trf',
+        ftol=EPS**0.5, xtol=EPS**0.5, gtol=EPS**0.5, loss='linear',
+        loss_scale=1.0, scaling=1.0, diff_step=None, tr_solver=None,
+        tr_options={}, jac_sparsity=None, max_nfev=None, verbose=0, args=(),
+        kwargs={}):
+    """Minimize the robustified sum of squares of nonlinear functions, subject
+    to bound constraints on independent variables.
```

Comment:
This should fit on a single line, because that's extracted into the summary table for ``optimize`` in the html refguide. Maybe something like ``Minimize the robustified sum of squares of nonlinear functions (with bounds)``?